### PR TITLE
tools: add eth tools support for DFL driver

### PR DESCRIPTION
**Pull request by anandaravuri**
FPGA N3000 production driver supports eth group ioctl’s to access eth group info , read and write registries, DFL driver doesn’t support any ioctl’s,  user can read/write eth group registers by enabling VFIO mdev.

- Enable FPGA DFL MDEV for Ethernet group device feature id 10, 
Steps to Enable eth group:
echo dfl-fme.0.8 > /sys/bus/dfl/drivers/dfl-eth-group/unbind
echo dfl-fme.0.7 > /sys/bus/dfl/drivers/dfl-eth-group/unbind
echo vfio-mdev-dfl > /sys/bus/dfl/devices/dfl-fme.0.7/driver_override
echo vfio-mdev-dfl > /sys/bus/dfl/devices/dfl-fme.0.8/driver_override
modprobe vfio_pci
modprobe vfio_iommu_type1
modprobe vfio_mdev
modprobe vfio_mdev_dfl
echo dfl-fme.0.7 >/sys/bus/dfl/drivers_probe
echo dfl-fme.0.8 >/sys/bus/dfl/drivers_probe
echo  83b8f4f2-509f-382f-3c1e-e6bfe0fa1001 > /sys/bus/dfl/devices/dfl-fme.0.7/mdev_supported_types/vfio-mdev-dfl-1/create
echo  83b8f4f2-509f-382f-3c1e-e6bfe0fa1002 > /sys/bus/dfl/devices/dfl-fme.0.8/mdev_supported_types/vfio-mdev-dfl-1/create

- Add  VFIO pybind 11 eth_group module to fpgadiag   for access Eth group vfio mdev.
 Eth group class has functions open, close, read reg and write reg functions:
  eth_group_open(): Opens Eth group VFIO mdev and MMP’s device feature MMIO region.
  eth_group_close(): Un map’s MMIO region and closes VFIO mdev.
  read_reg(): Implements ETH GROUP MailBox commination, writes Address to eth group control register and  Reads data using ETH GROUP status registers.
 Write_reg(): Implements ETH GROUP MailBox communication , writes Address to eth group control register and  Reads data valid using ETH GROUP status registers.
 mac_reset(): resets MAC Config

- Added Setup.y to install Eth group pybind11 module  and updated opae.spec.in

- Modified common.py, fecmode.py, fpgalpbk.py, fpgamac.py, fpgastats.py, fvlbypass.py, mactest.py to read/write registers using eth group pubind11 module.

### CODE REVIEW

File: scripts/enable_eth_group_mdev.sh [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+FPGA_FEATURID=/sys/bus/pci/devices/*$bus_num*/fpga_region/region*/dfl-fme*/dfl-fme*/feature_id
+echo " Feature: $FPGA_FEATURID"
+feature_list=()
+for file in $FPGA_FEATURID
+do
+    echo "Feaure id $file"
+    feature_id=$(cat $file)
+    echo "Feature id: $feature_id"
+	if [[ $feature_id != 0x10 ]]; then
```
Comment (tswhison) [inline scripts/enable_eth_group_mdev.sh:40]: What does the 0x10 stand for?  Will it ever change?

File: scripts/enable_eth_group_mdev.sh [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	
+done
+
+
+modprobe vfio_pci
+modprobe vfio_iommu_type1
+modprobe vfio_mdev
+modprobe vfio_mdev_dfl
+var=0
+gudi_array=("83b8f4f2-509f-382f-3c1e-e6bfe0fa1001" "83b8f4f2-509f-382f-3c1e-e6bfe0fa1002" "83b8f4f2-509f-382f-3c1e-e6bfe0fa1003")
```
Comment (tswhison) [inline scripts/enable_eth_group_mdev.sh:55]: s/gudi_array/guid_array/
why 3 GUIDs?

File: tools/extra/fpgadiag/eth_group.cpp [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+#define CMD_WR			2
+#define SELECT_IP		0
+#define SELECT_FEAT		1
+#define ETH_GROUP_STAT		0x18
+#define ETH_GROUP_SELECT_FEAT	(1 << 0)
+#define ETH_GROUP_PHY		1
+#define ETH_GROUP_MAC		2
+#define ETH_GROUP_ETHER		3
+
+#define DVE_VFIO_PATH  "/dev/vfio/vfio"
```
Comment (tswhison) [inline tools/extra/fpgadiag/eth_group.cpp:48]: s/DVE_VFIO_PATH/DEV_VFIO_PATH/

File: tools/extra/fpgadiag/eth_group.cpp [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+
+	eth_ctl.eth_cmd = CMD_RD;
+	eth_ctl.ctl_addr = addr;
+	eth_ctl.ctl_fev_select = flags & ETH_GROUP_SELECT_FEAT;
+
+	// write to ctrl reg
+	*((volatile uint64_t *)(mmap_ptr + ETH_GROUP_CTRL))
+		= (uint64_t)eth_ctl.csr;
+
+	//read untill status reg bit valid
```
Comment (tswhison) [inline tools/extra/fpgadiag/eth_group.cpp:271]: s/untill/until/

File: tools/extra/fpgadiag/eth_group.cpp [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+	eth_ctl.eth_cmd = CMD_WR;
+	eth_ctl.ctl_addr = addr;
+	eth_ctl.ctl_data = data;
+	eth_ctl.ctl_fev_select = flags & ETH_GROUP_SELECT_FEAT;
+
+	// write to ctrl reg
+	*((volatile uint64_t *)(mmap_ptr + ETH_GROUP_CTRL))
+		= (uint64_t)eth_ctl.csr;
+
+	//read untill status reg bit valid
```
Comment (tswhison) [inline tools/extra/fpgadiag/eth_group.cpp:323]: s/untill/until/

File: scripts/enable_eth_group_mdev.sh (hunk 2 above)
Comment (anandaravuri) [reply scripts/enable_eth_group_mdev.sh]: fixed

File: tools/extra/fpgadiag/eth_group.cpp (hunk 3 above)
Comment (anandaravuri) [reply tools/extra/fpgadiag/eth_group.cpp]: fixed

File: tools/extra/fpgadiag/eth_group.cpp (hunk 4 above)
Comment (anandaravuri) [reply tools/extra/fpgadiag/eth_group.cpp]: fixed

File: tools/extra/fpgadiag/eth_group.cpp (hunk 5 above)
Comment (anandaravuri) [reply tools/extra/fpgadiag/eth_group.cpp]: fixed

File: scripts/enable_eth_group_mdev.sh (hunk 2 above)
Comment (anandaravuri) [reply scripts/enable_eth_group_mdev.sh]: Two guid array, 
Host guid "83b8f4f2-509f-382f-3c1e-e6bfe0fa1001" 
Ethernet guid "83b8f4f2-509f-382f-3c1e-e6bfe0fa1002"

File: scripts/enable_eth_group_mdev.sh (hunk 1 above)
Comment (anandaravuri) [reply scripts/enable_eth_group_mdev.sh]: Eth group DFH feature ID for FPGA N3000

**Review by tswhison — APPROVED** [review]

**Review by asgardkm — APPROVED** [review]

**Pull request merged**